bip-annex.mediawiki

```diff
@@ -0,0 +1,173 @@
+defined to make accessible annex fields to Script operations. </ref> Beyond, the annex format aims to
+be reusable across spends of SegWit versions.
+
+== Specification ==
+
+=== CompressedInt Integer Encoding ===
+
+Variable-length integers: bytes are a MSB base-128 encoding of the number.
+The high bit in each byte signifies whether another digit follows. To make
+sure the encoding is one-to-one, one is subtracted from all but the last digit.
+Thus, the byte sequence a[] with length len, where all but the last byte
+has bit 128 set, encodes the number:
+
+  (a[len-1] & 0x7F) + sum(i=1..len-1, 128^i*((a[len-i-1] & 0x7F)+1))
+
+Properties:
+ * Very small (0-127: 1 byte, 128-16511: 2 bytes, 16512-2113663: 3 bytes)
+ * Every integer has exactly one encoding
+ * Encoding does not depend on size of original integer type
+ * No redundancy: every (infinite) byte sequence corresponds to a list
+   of encoded integers.
+
+Examples:
+ * 0:         [0x00]  256:        [0x81 0x00]
+ * 1:         [0x01]  16383:      [0xFE 0x7F]
+ * 127:       [0x7F]  16384:      [0xFF 0x00]
+ * 128:  [0x80 0x00]  16511:      [0xFF 0x7F]
+ * 255:  [0x80 0x7F]  65535: [0x82 0xFE 0x7F]
+ * 2^32:           [0x8E 0xFE 0xFE 0xFF 0x00]
+
+ read_CompressedInt():
+     result = 0
+     while not eof():
+         b = read_bytes(1)
+         if b < 128: return result + b
+         result += b - 127
+         result *= 128
+     fail()
+
+ write_CompressedInt(n):
+     out = []
+     while True:
+         out.append( n % 128 )
+         if n <= 127: break
+         n = (n // 128) - 1
+     while len(out) > 1:
+         write(out.pop() | 0x80)
+     write(out.pop())
+
+=== Type-Length-Value Format ===
+
+The annex is defined as containing an ordered set of "type, value" pairs, where the type is a non-negative integer, and the value is a byte stream, and the pairs are listed in non-decreasing order by type.
+
+The annex is encoded as follows:
+
+ write(0x50)
+ last_type = 0
+ for type, value in annex:
+    delta = type - last_type
+    assert delta >= 0, "annex must be ordered by type"
+    if length(value) < 127:
+        write_CompressedInt(delta * 128 + length(value))
+    else:
+        write_CompressedInt(delta * 128 + 127)
+        write_CompressedInt(length(value) - 127)
+    write(value)
+    last_type = type
+
+And conversely the annex may be decoded as follows:
+
+ assert read_bytes(1) == 0x50, "annex must begin with annex marker"
+ last_type = 0
+ annex = []
+ while not eof():
+    deltalen = read_CompressedInt()
+    type = last_type + (deltalen >> 7)
+    length = deltalen & 0x7F
+    if length == 0x7F:
+        length += read_CompressedInt()
+    value = read_bytes(length)
+    annex.append( (type, value) )
+    last_type = type
+
+Rather than encoding the type directly, we encode the difference between
+the previous type (initially 0), both minimising the encoding and ensuring
+a canonical ordering for annex entries.
```

Comment:
canonical up to reordering of entries of the same type.

Reply:
The way I think about the annex is that as a total function from non-negative integers to vectors of byte strings.

So if you're mapping type 9 to the vector `["abc", "def"]` then that is distinct from mapping type 9 to the vector `["def", "abc"]`, and "canonical" here means there's exactly one encoding for each mapping.

In particular, I'm thinking that a script opcode to examine an annex entry might look like `9 PUSH_ANNEX 2 EQUALVERIFY` -- now you have "abc" at the top of the stack and "def" beneath it. In that case the annex entries of a single type cannot always be reordered without changing their semantics.

(The current text says `The annex is defined as containing an ordered set of "type, value" pairs,` -- my thinking was that "ordered set" already captures the idea that reordering entries changes the semantics)